File: apps/api/app/hashtags.py
from __future__ import annotations

import re

TOKEN = re.compile(r"^[a-z0-9-]{1,24}$")
SPLIT = re.compile(r"[\s,]+")


def parse_hashtags(raw: str, *, limit: int = 8) -> list[str]:
    parts = SPLIT.split((raw or "").strip().lower())
    out: list[str] = []
    seen: set[str] = set()
    for part in parts:
        tag = part.lstrip("#")
        if not tag or tag in seen or not TOKEN.fullmatch(tag):
            continue
        seen.add(tag)
        out.append(tag)
        if len(out) >= limit:
            break
    return out

File: apps/web/src/lib/hashtags.ts
const TOKEN = /^[a-z0-9-]{1,24}$/;

export function parseHashtags(raw: string, limit = 8): string[] {
  const parts = raw
    .trim()
    .toLowerCase()
    .split(/[\s,]+/)
    .map((part) => part.replace(/^#/, ""))
    .filter(Boolean);
  const out: string[] = [];
  const seen = new Set<string>();
  for (const tag of parts) {
    if (seen.has(tag) || !TOKEN.test(tag)) continue;
    seen.add(tag);
    out.push(tag);
    if (out.length >= limit) break;
  }
  return out;
}

export function captionTags(tags: string[] | undefined): string[] {
  return (tags ?? []).filter((tag) => tag && tag !== "seed");
}
